src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,323 @@
+
+    // Iterate on roots
+    while (t_lower < tmax_)
+    {
+        // Set bounds on current iteration
+
+        t_alpha = t_lower;
+        t_beta = t_upper;
+
+        // Find value in root function
+        ft_alpha = root(t_lower, x, y, u, v, r_b_, a_);
+        ft_beta = root(t_upper, x, y, u, v, r_b_, a_);
+
+        // If bounds exceed tmax break
+        if (t_lower > tmax_)
+        {
+            break;
+        }
+
+        // Only iterate when roots have different signs
+        if ((0 < ft_alpha) - (ft_alpha < 0) != (0 < ft_beta) - (ft_beta < 0))
+        {
+            // Regula Falsi Iteration
+            ft_gamma = 1;
+            while (std::fabs(ft_gamma) >= tol_conv)
+            {
+                // Iterate on root
+                t_gamma = (t_alpha * ft_beta - t_beta * ft_alpha)
+                          / (ft_beta - ft_alpha);
+
+                // Obtain root value of iterated root
+                ft_gamma = root(t_gamma, x, y, u, v, r_b_, a_);
+
+                // Update bounds with iterated root
+                if ((0 < ft_beta) - (ft_beta < 0)
+                    == (0 < ft_gamma) - (ft_gamma < 0))
+                {
+                    t_beta = t_gamma;
+                    ft_beta = ft_gamma;
+                }
+                else
+                {
+                    t_alpha = t_gamma;
+                    ft_alpha = ft_gamma;
+                }
+            }
+            // Extract root and calculate point on involute
+            angle = t_gamma + a_;
+            x_inv = r_b_ * (std::cos(angle) + t_gamma * std::sin(angle));
+            y_inv = r_b_ * (std::sin(angle) - t_gamma * std::cos(angle));
+
+            // Check if point is interval
+            if (std::fabs(t_gamma) >= std::fabs(tmin_) - tol_point
+                && std::fabs(t_gamma) <= std::fabs(tmax_))
+            {
+                // Obatin direction to point on Involute
+                real_type u2 = x_inv - x;
+                real_type v2 = y_inv - y;
+
+                // Dot with direction of particle
+                real_type dot = u * u2 + v * v2;
+                // Obtain distance to point
+                real_type newdist = std::sqrt(ipow<2>(u2) + ipow<2>(v2));
+                // Only record distance if dot product is positive
+                if (dot >= 0 && newdist > tol_point)
+                {
+                    dist[j] = newdist;
+                    j++;
+                }
+            }
+
+            t_lower = t_upper;
+            t_upper += pi;
+        }
+        else
+        {
+            // Incremet interval slowly until root is in interval
+            // i used to slow down approach to next root
+            t_lower = t_upper;
+            t_upper += pi / i;
+            i++;
+        }
+
+        for (int k = 0; k < j; k++)
+        {
+            // Convert 2-D distances to 3-D distances
+            result[k] = dist[k] * convert;
+        }
+    }
+    return result;
+}
```

Comment:
This function is far too long and complicated: it has 200 lines and 13 different conditionals (if/while/for) in it. Can you think of ways to break it into smaller *testable* pieces, each of which has only a couple of conditionals?

Reply:
Got it down to 120 lines and 7 conditionals (counting one else) in the main function